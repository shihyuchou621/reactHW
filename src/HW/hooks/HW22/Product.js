import React, { createRef }  from 'react';

import style from './style.module.css';

export default function Product(props) {
  const DOM = createRef();
  const handleSubmit = e => {
    e.preventDefault();
    props.onSubmit(+DOM.current.value + props.amount);
    DOM.current.value = "";
  };

  const {
    name,
    price,
  } = props;

  return (
    <div className={style.product}>
      <div>{name}</div>
      <div>${price}</div>
      <form
        data-testid="form"
        name={name}
        onSubmit={handleSubmit}
      >
        <input
          data-testid="input"
          className={style.input}
          name={name}
          ref={DOM}
          type="number"
          placeholder="數量"
          min="0"
          max="99"
        />
        <button
          className="btn btn-outline-primary btn-sm"
        >加入購物車</button>
      </form>
    </div>
  );
}
