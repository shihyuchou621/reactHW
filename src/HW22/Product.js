import React, { Component } from 'react'

import './style.css'

export default class Product extends Component {
  handleSubmit = e => {
    e.preventDefault();
    this.props.onSubmit(this.props.index, +this.productInput.value)
    this.productInput.value = "";
  }
  
  render() {
    const {
      name,
      price,
    } = this.props;

    return (
      <div className="product">
        <div>{name}</div>
        <div>${price}</div>
        <form onSubmit={this.handleSubmit}>
          <input className="input"
            ref={element => this.productInput = element}
            type="number"
            placeholder="數量"
            min="0" 
            max="99" 
          />
          <button className="btn btn-outline-primary btn-sm">加入購物車</button>
        </form>
      </div>
    )
  }
}
