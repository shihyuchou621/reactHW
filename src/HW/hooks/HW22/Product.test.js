import React from 'react';
import { render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Product from './Product';

afterEach(cleanup);

test('should have correct value', () => {
  const onSubmit = jest.fn(
    (index, value) => ({index, value})
  );
  const { getByTestId, container } = render(
    <Product
      name="橘子"
      price={20}
      index={1}
      onSubmit={onSubmit}
    />
  );

  userEvent.type(getByTestId('input'), '5{enter}');
  expect(container).toMatchSnapshot();

  userEvent.type(getByTestId('input'), '2{enter}');
  expect(container).toMatchSnapshot();

});
