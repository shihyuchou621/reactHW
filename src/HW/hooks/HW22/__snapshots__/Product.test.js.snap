// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have correct value 1`] = `
<div>
  <div
    class="product"
  >
    <div>
      橘子
    </div>
    <div>
      $
      20
    </div>
    <form
      data-testid="form"
      name="橘子"
    >
      <input
        class="input"
        data-testid="input"
        max="99"
        min="0"
        name="橘子"
        placeholder="數量"
        type="number"
      />
      <button
        class="btn btn-outline-primary btn-sm"
      >
        加入購物車
      </button>
    </form>
  </div>
</div>
`;

exports[`should have correct value 2`] = `
<div>
  <div
    class="product"
  >
    <div>
      橘子
    </div>
    <div>
      $
      20
    </div>
    <form
      data-testid="form"
      name="橘子"
    >
      <input
        class="input"
        data-testid="input"
        max="99"
        min="0"
        name="橘子"
        placeholder="數量"
        type="number"
      />
      <button
        class="btn btn-outline-primary btn-sm"
      >
        加入購物車
      </button>
    </form>
  </div>
</div>
`;
